graph/coloring/coloring.go

```diff
@@ -0,0 +1,785 @@
+// may be used as a Terminator.
+type Terminator interface {
+	// Done returns a channel that is closed when work
+	// should be terminated. Done may return nil if this
+	// work can never be canceled.
+	// Successive calls to Done should return the same value.
+	Done() <-chan struct{}
+
+	// If Done is not yet closed, Err returns nil.
+	// If Done is closed, Err returns a non-nil error
+	// explaining why.
+	// After Err returns a non-nil error, successive
+	// calls to Err should return the same error.
+	Err() error
+}
+
+// DsaturExact returns the exact minimal chromatic number of g and a
+// corresponding vertex coloring using the branch-and-bound Dsatur coloring
+// algorithm of Brélaz. If the provided terminator is cancelled or times out
+// before completion, the terminator's reason for termination will be returned
+// along with a potentially sub-optimal chromatic number and coloring. If
+// term is nil, DsaturExact will run to completion.
+// See Brélaz doi:10.1145/359094.359101 for details of the algorithm.
+func DsaturExact(term Terminator, g graph.Undirected) (k int, colors map[int64]int, err error) {
+	// This is implemented essentially as described in algorithm 1 of
+	// doi:10.1002/net.21716 with the exception that we obtain a
+	// tighter upper bound by doing a single run of an approximate
+	// Brélaz Dsatur coloring, using the result if the recurrence is
+	// cancelled.
+
+	nodes := g.Nodes()
+	n := nodes.Len()
+	if n == 0 {
+		return
+	}
+
+	lb, maxClique := maximumClique(g)
+	if lb == n {
+		return lb, colorClique(maxClique), nil
+	}
+
+	order := bySaturationDegree(nodes, g, make(map[int64]int))
+	order.heuristic = order.dsatur
+	// Find initial coloring via Dsatur heuristic.
+	ub, initial := greedyColoringOf(g, order, order.colors)
+	if lb == ub {
+		return ub, initial, nil
+	}
+
+	selector := &order.saturationDegree
+	cand := newDsaturColoring(order.nodes, make(map[int64]int))
+	k, colors, err = dSaturExact(term, selector, cand, 0, ub, nil)
+	if colors == nil {
+		return ub, initial, err
+	}
+	if k == lb {
+		err = nil
+	}
+	return k, colors, err
+}
+
+// dSaturColoring is a partial graph coloring.
+type dSaturColoring struct {
+	colors    map[int64]int
+	uncolored set.Int64s
+}
+
+// newDsaturColoring returns a dSaturColoring representing a partial coloring
+// of a graph with the given nodes and colors.
+func newDsaturColoring(nodes []graph.Node, colors map[int64]int) dSaturColoring {
+	uncolored := make(set.Int64s)
+	for _, v := range nodes {
+		vid := v.ID()
+		if _, ok := colors[vid]; !ok {
+			uncolored.Add(vid)
+		}
+	}
+	return dSaturColoring{
+		colors:    colors,
+		uncolored: uncolored,
+	}
+}
+
+// color moves a node from the uncolored set to the colored set.
+func (c dSaturColoring) color(id int64) {
+	if !c.uncolored.Has(id) {
+		if _, ok := c.colors[id]; ok {
+			panic("coloring: coloring already colored node")
+		}
+		panic("coloring: coloring non-existent node")
+	}
+	c.uncolored.Remove(id)
```

Comment:
Is it intentional that `id` is not added to `c.colors` here?

Reply:
Yes. The colour of the node is not yet known, so while we know that it is coloured, the actual colour that it is won't be resolved until `dSaturExact` has returned (the assignments happen in the feasible colour loop of that function). I can add some text above this line to make that clear.

```
	// The node is marked as uncolored, but is not
	// validly colored until the dSaturExact caller
	// has completed its recursive exploration of the
	// feasible colors. So don't explicity color it yet.
```

Reply:
Yes please, that comment would help. Thanks.

Reply:
Dome

Reply:
Now with the correct rather than diametrically wrong wording.